src/deepsparse/sentence_transformers/sentence_transformer.py

```diff
@@ -0,0 +1,182 @@
+
+import torch
+from optimum.deepsparse import DeepSparseModelForFeatureExtraction
+
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_MODEL_NAME = "zeroshot/bge-small-en-v1.5-quant"
+
+
+class SentenceTransformer:
+    def __init__(
+        self,
+        model_name_or_path: str = DEFAULT_MODEL_NAME,
+        export: bool = False,
+        max_seq_length: int = 512,
+    ):
+
+        self.model_name_or_path = model_name_or_path
+        self.model = DeepSparseModelForFeatureExtraction.from_pretrained(
+            model_name_or_path, export=export
+        )
+        self.tokenizer = get_preprocessor(model_name_or_path)
+
+        self._max_seq_length = max_seq_length
+        self._batch_size = 1
+
+    def encode(
+        self,
+        sentences: Union[str, List[str]],
+        batch_size: int = 1,
+        show_progress_bar: bool = None,
+        output_value: str = "sentence_embedding",
+        convert_to_numpy: bool = True,
+        convert_to_tensor: bool = False,
+        normalize_embeddings: bool = False,
+    ) -> Union[List[torch.Tensor], np.ndarray, torch.Tensor]:
+
+        # TODO: support executing with batch size > 1
+        batch_size = 1
+
+        if show_progress_bar is None:
+            show_progress_bar = (
+                logger.getEffectiveLevel() == logging.INFO
+                or logger.getEffectiveLevel() == logging.DEBUG
+            )
+
+        if convert_to_tensor:
+            convert_to_numpy = False
+
+        if output_value != "sentence_embedding":
+            convert_to_tensor = False
+            convert_to_numpy = False
+
+        input_was_string = False
+        if isinstance(sentences, str) or not hasattr(
+            sentences, "__len__"
+        ):  # Cast an individual sentence to a list with length 1
+            sentences = [sentences]
+            input_was_string = True
+
+        all_embeddings = []
+        length_sorted_idx = np.argsort([-self._text_length(sen) for sen in sentences])
+        sentences_sorted = [sentences[idx] for idx in length_sorted_idx]
+
+        for start_index in trange(
+            0, len(sentences), batch_size, desc="Batches", disable=not show_progress_bar
+        ):
+            sentences_batch = sentences_sorted[start_index : start_index + batch_size]
+
+            model_inputs = self.tokenize(sentences_batch)
+            model_output = self.model(**model_inputs)
+
+            out_features = {}
+            out_features["sentence_embedding"] = self.mean_pooling(
+                model_output, model_inputs["attention_mask"]
+            )
+
+            if output_value == "token_embeddings":
+                embeddings = []
+                for token_emb, attention in zip(
+                    out_features[output_value], out_features["attention_mask"]
+                ):
+                    last_mask_id = len(attention) - 1
+                    while last_mask_id > 0 and attention[last_mask_id].item() == 0:
+                        last_mask_id -= 1
```

Comment:
is attention mask gauranteed to be 1s followed by 0s? could remove this while loop with attention.sum() and a slice, otherwise, could select from `token_emb` by `attention.nonzero()`

Reply:
This part was copied from SentenceTransformers, so good call on simplifying this. I believe in the guarantee so I will choose the first implementation